crates/re_query_cache/src/flat_vec_deque.rs

```diff
@@ -0,0 +1,789 @@
+        let deque = Self {
+            values: values.collect(),
+            offsets: std::iter::once(num_values).collect(),
+        };
+        self.insert_with(entry_index, deque);
+    }
+
+    /// Prepends multiple entries, each comprised of the multiple elements given in `entries`,
+    /// to the deque.
+    ///
+    /// This is the same as `self.insert_range(0, entries)`.
+    ///
+    /// See [`Self::insert_range`] for more information.
+    #[inline]
+    pub fn push_range_front(&mut self, entries: impl IntoIterator<Item = Vec<T>>) {
+        self.insert_range(0, entries);
+    }
+
+    /// Appends multiple entries, each comprised of the multiple elements given in `entries`,
+    /// to the deque.
+    ///
+    /// This is the same as `self.insert_range(self.num_entries(), entries)`.
+    ///
+    /// See [`Self::insert_range`] for more information.
+    #[inline]
+    pub fn push_range_back(&mut self, entries: impl IntoIterator<Item = Vec<T>>) {
+        self.insert_range(self.num_entries(), entries);
+    }
+
+    /// Inserts multiple entries, starting at `entry_index` onwards, each comprised of the multiple elements
+    /// given in `entries`.
+    ///
+    /// This is O(1) if `entry_index` corresponds to either the start or the end of the deque.
+    /// Otherwise, this requires splitting the deque into two pieces then stitching them back together
+    /// at both ends of the added data.
+    ///
+    /// Panics if `entry_index` is out of bounds.
+    /// Panics if any of the value arrays in `entries` is empty.
+    #[inline]
+    pub fn insert_range(&mut self, entry_index: usize, entries: impl IntoIterator<Item = Vec<T>>) {
+        let deque = Self::from_vecs(entries);
+        self.insert_with(entry_index, deque);
+    }
+
+    /// Prepends another full deque to the deque.
+    ///
+    /// This is the same as `self.insert_with(0, rhs)`.
+    ///
+    /// See [`Self::insert_with`] for more information.
+    #[inline]
+    pub fn push_front_with(&mut self, rhs: FlatVecDeque<T>) {
+        self.insert_with(0, rhs);
+    }
+
+    /// Appends another full deque to the deque.
+    ///
+    /// This is the same as `self.insert_with(0, rhs)`.
+    ///
+    /// See [`Self::insert_with`] for more information.
+    #[inline]
+    pub fn push_back_with(&mut self, rhs: FlatVecDeque<T>) {
+        self.insert_with(self.num_entries(), rhs);
+    }
+
+    /// Inserts another full deque, starting at `entry_index` and onwards.
+    ///
+    /// This is O(1) if `entry_index` corresponds to either the start or the end of the deque.
+    /// Otherwise, this requires splitting the deque into two pieces then stitching them back together
+    /// at both ends of the added data.
+    ///
+    /// Panics if `entry_index` is out of bounds.
+    /// Panics if any of the value arrays in `entries` is empty.
+    pub fn insert_with(&mut self, entry_index: usize, mut rhs: FlatVecDeque<T>) {
+        if entry_index == self.num_entries() {
```

Comment:
same off-by-one bug here

```suggestion
        if entry_index + 1 == self.num_entries() {
```

Reply:
That's on purpose, we want to insert _beyond_ the last element.

Reply:
oh right 🤦